src/realm/object-store/sync/sync_session.cpp

```diff
@@ -1536,3 +1595,29 @@ util::Future<std::string> SyncSession::send_test_command(std::string body)
 
     return m_session->send_test_command(std::move(body));
 }
+
+void SyncSession::make_active_subscription_set()
+{
+    util::CheckedUniqueLock lock(m_state_mutex);
+
+    if (!m_active_subscriptions_after_migration)
+        return;
+
+    REALM_ASSERT(m_flx_subscription_store);
+
+    // Create subscription set from the subscriptions used to download the fresh realm after migration.
+    auto active_mut_sub = m_flx_subscription_store->get_active().make_mutable_copy();
+    active_mut_sub.import(*m_active_subscriptions_after_migration);
+    lock.unlock();
+
+    // Commit without holding the lock to prevent a deadlock.
+    auto active_sub = active_mut_sub.commit();
+
+    lock.lock();
+    // Mark the subscription set complete.
+    auto sub = m_flx_subscription_store->get_mutable_by_version(active_sub.version());
+    sub.update_state(sync::SubscriptionSet::State::Complete);
+    sub.commit();
```

Comment:
I don't think we can do all in one commit unfortunately. If I understand the code at https://github.com/realm/realm-core/blob/master/src/realm/sync/subscriptions.cpp#L535-L555 correctly, the subscription set has to be in the Uncommitted state for the subscriptions to actually be created. We need that otherwise the write validator will reject every write.

Reply:
@jbreams You are right, we can do it in one commit. We have two variables to track the state, m_old_state and m_state which confused me a bit. m_old_state is always Uncommitted so the subscriptions are created, and because m_state is updated to Complete, on_new_subscription_set does not fire so no need to release the lock.